### ISSUE TITLE
AzureBlobCache plugin use redirect instead to rewrite

### ISSUE THREAD

**Issue by abadyl** [JimBobSquarePants/ImageProcessor#161]
Hi, just installed AzureBlobCache plugin. It work well, but with one remark.

When I access images, I all time redirected to CDN url. That mean if my image already in cache, server will redirect me to Azure CDN url. 

This is extra round trip, and also issue with SEO for images.

Is it possible to not use redirect, but make call for image to azure blob via API, get image stream, and then copy it to output stream of response?

In this case we will have clean SEO for images, and without extra redirect.

What was an idea to use redirect?

Thanks,

**Comment by JimBobSquarePants** [comment]
Hi there,

Could you clarify in what way this would be a SEO issue?

If I were to request the urls directly from the CDN Google would treat that in the same way as my 302 redirect from the request url. (I use 302 because if the source file is updated with the same name, the request url could be the same but the CDN url would be different.)

There's no extra round trip; the url is intercepted and redirected. I have to use a url hosted within the currently running application for all image requests because there is no way to intercept an outgoing request to another domain.

**Comment by abadyl** [comment]
Hi, 
I agree that 302 redirect will work from SEO point.

But, if you have on a page i.e. 100 images => browser will be redirected 100 times. I also understand that if we will user real pull CDN it is not an a problem, but if we not use CDN at all. Or if your container should be private?

It also difficult to tell  how it (redirects) will influence on overall performance of website.

Why you not use API or webrequest to get cache stream from server and write it to response?

Will code bellow make sence?

        
        public void RewritePath(HttpContext context)
        {
            HttpWebRequest httpWebRequest = (HttpWebRequest)WebRequest.Create(this.cachedRewritePath);
            httpWebRequest.Method = "HEAD";

            using (HttpWebResponse httpWebResponse = (HttpWebResponse)httpWebRequest.GetResponse())
            {
                //get image stream from cloud
                Stream ourStram = httpWebResponse.GetResponseStream();

                ourStram.CopyTo(context.Response.OutputStream);
            }
        }

**Comment by JimBobSquarePants** [comment]
Interesting... Would you have the opportunity to fork and test this? 

I'm incredibly busy just now so you would most probably be faster than me.

**Comment by abadyl** [comment]
Sure, this weekend I will try to play with it.

Will keep you informed.

**Comment by JimBobSquarePants** [comment]
Brilliant! Thanks for your input :smile:

**Comment by abadyl** [comment]
Hi, 
This code example are working well. At least I was able to see my image in browser without redirect.
You can see my fork (https://github.com/abadyl/ImageProcessor).

**Comment by robertjf** [comment]
Just my 2 cents from a lurker - to me, this seems like it's going to add extra overhead and possible lag - it seems to me that you're effectively copying the contents of a file from one server (the cache) to another (your server) and then shuffling them to the client - wouldn't it be more efficient for the client to be told to pull the contents directly from the caching server?

**Comment by abadyl** [comment]
Yes, overhead will be in compare to redirect.

But,
1. If you use CDN, then this is not issue at all (the same for redirect)
2. In my example it is more easy configure caching policy (expiration time on CDN). So you have more control over it. If you use Redirect, you will  need update caching settings of each blob after it need to be changed.
3. When you use default disk cache, it is also transfer (copy) from file stream to network stream.
4. What, if your container will be private. Then you just can use API to get stream instead of GET request.

To summarize, 

Both variants will work. Both have pros and cons. 
In our project point 2 is very important, this is why I am looking for alternative solution.

It is also option to have 2 plugins, one with redirect and one with stream copy (like proxy).

**Comment by cosmo0** [comment]
Please note that replacing the 302 with the stream will switch the burden of downloading the image from the client to the server. I don't think it's a great idea. The original question si basically "why are you using the power of a cdn to its full potential ?".

**Comment by abadyl** [comment]
Yes, this is clear. And with CDN enabled it will be done only first time. All other requests for the same image will be loaded from CDN.

I.E.
You can use redirect when you don't have CDN enabled
or You can use stream copy when you have CDN enabled and need more control over blobs  (for more easy cache control/expiration management, security management)

P.S. I also not fully sure that most popular search engines treat 302 redirect correctly (as 302, but not as 301).

Goal of created issue not to replace redirect, but have alternative variant for serving cached images. 

If speak about real example, in blobs we already have more than 12000 images. When CDN will be enabled we need to be sure that cache control set up correctly. So, we don't want to go through all images, we want to control it in one place. As result variant with redirect not work well for us.

**Comment by cosmo0** [comment]
I'm sorry, I'm not an expert on CDN or SOE (I've more worked on Windows apps lately), but I thought the following:

* the interest of a CDN are:
  * to minimize the bandwidth on your servers, and instead host static content on other servers
  * to provide users with a very-wide-bandwidth server to download their contents (the CDN being optimized for BW)
  * to allow the client to cache locally their image, in case they happen to load the same file from a different website, like the Jquery files from Google's CDN (obviously not very relevant here)
* the main search engines (Google, Yahoo, Bing) are perfectly capable of handling properly any kind of redirect, and in fact probably treat redirects differently than one would assume
* the SEO impact of an image is pretty low, or even nil, because it's more the alt and surrounding text that are used to describe it (I'm not very sure about that, correct me if I'm wrong)

All in all, I have trouble understanding the point, and the benefits, of such a feature. Obviously, if it's an option, there is no real harm (except to the project's complexity), but I'm not sure of the interest.

**Comment by cosmo0** [comment]
@abadyl you have finished your work on your fork ? Have you made a pull request ? It seems you're not the only one interested in this feature (see #178)

PS: I have since understood the interest of this feature, since @gnpretorius has explained it to me :smile: (basically, I thought it worked inversely of what it does).

**Comment by abadyl** [comment]
Hi, 

It is now only in my fork. I did't create pull request because I wasn't sure (based on comments above) if it is nice feature to be integrated to imageprocessor.

So, will send pull request this week,

Best regards.
Alexey

**Comment by cosmo0** [comment]
Thanks!

I really didn't understand the point of your modifications, until the explanation at #178 .

I'd like @gnpretorius to check out your PR and see if it solves his problem.

**Comment by gnpretorius** [comment]
Hi @abadyl 

The [code](https://github.com/abadyl/ImageProcessor/commit/ddc13659fae56a15c34b6af6ec960ef37462610f?diff=split) here looks good. Should we make this configurable? Some people might want to keep the redirect behaviour?

@cosmo0, let me know your thoughts?

G

**Comment by cosmo0** [comment]
I'm sure setting this as an option would be awesome, if it solves both your problems :smile:

**Comment by JimBobSquarePants** [comment]
:+1: on optional. Add it to the cache settings. Btw loving the interest and collaboration!

**Comment by JimBobSquarePants** [comment]
Added in version [1.0.4](https://www.nuget.org/packages/ImageProcessor.Web.Plugins.AzureBlobCache/1.0.4) 

I've added the extra configuration to the docs to show how to switch to the new behaviour. 

http://imageprocessor.org/imageprocessor-web/plugins/azure-blob-cache/#config

Thanks one and all for your persistence and contribution. It is is very much appreciated :+1:

**Comment by olaj** [comment]
I was not sure if i should ask here or Stack Overflow but i posted a question about this on Stack Overflow and well nobody has answered so i thought i might ask here :)

http://stackoverflow.com/questions/35550998/optimal-settings-for-serving-images-with-imageprocessor-azureblobcache

Maybe someone else need more clarification!

Thanks for a great framework!

**Comment by leye0** [comment]
I feel like there is no good solution for now. The main problem is the SEO issue mentionned by the original poster @abadyl. Who says SEO says "the only solution is to maintain a list of UserAgent string". Because the rest of the ImageProcessor/AzureBlobStorage stream is optimal and should not be modified.
